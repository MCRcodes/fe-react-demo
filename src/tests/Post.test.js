import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import Post from "../components/Post";

describe("Post", () => {
  const validProps = {
    postData: {
      author: "test author",
      body: "test body",
      date: "test date",
      isPublished: true,
      tags: ["test tag1", "test tag2", "test tag3"],
      title: "test title",
    },
    handleUpvote: jest.fn(),
  };

  test("renders as expected", () => {
    const testRenderer = renderer
      .create(
        <Post
          postData={validProps.postData}
          handleUpvote={validProps.handleUpvote}
        />
      )
      .toJSON();

    expect(testRenderer).toMatchSnapshot();
  });

  test("Post author renders correctly", () => {
    const { getByText } = render(
      <Post
        postData={validProps.postData}
        handleUpvote={validProps.handleUpvote}
      />
    );

    expect(getByText("Author: test author")).toBeInTheDocument;
  });

  test("Renders upvote button with expected value", () => {
    render(
      <Post
        postData={validProps.postData}
        handleUpvote={validProps.handleUpvote}
      />
    );
    const buttonElement = screen.getByText(/upvote this/i);

    expect(buttonElement).toBeInTheDocument();
  });

  test("list renders correct number of items", () => {
    const { getAllByRole } = render(
      <Post
        postData={validProps.postData}
        handleUpvote={validProps.handleUpvote}
      />
    );
    const listItems = getAllByRole('listitem');

    expect(listItems.length).toBe(3);
  });

  test("register upvote button clicks", () => {
    const { getByText } = render(
      <Post
        postData={validProps.postData}
        handleUpvote={validProps.handleUpvote}
      />
    );

    fireEvent.click(getByText("Upvote this"));
    expect(validProps.handleUpvote).toBeCalledTimes(1);
  });
});
