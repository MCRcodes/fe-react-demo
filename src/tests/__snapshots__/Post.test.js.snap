// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders as expected 1`] = `
<div
  className="post"
>
  <div
    className="post-body"
  >
    <h2>
      test title
    </h2>
    <p>
      test body
    </p>
  </div>
  <div
    className="post-counter"
  >
    <span
      style={
        Object {
          "marginRight": "12px",
        }
      }
    >
      Total votes: 
      0
    </span>
    <button
      onClick={[Function]}
      type="button"
      value="test title"
    >
      Upvote this
    </button>
  </div>
  <div
    className="post-author"
  >
    Author: 
    test author
  </div>
  <div
    className="post-date"
  >
    Published: 
    test date
  </div>
  <h3>
    Tags:
  </h3>
  <ul
    className="post-tags"
  >
    <li>
      test tag1
    </li>
    <li>
      test tag2
    </li>
    <li>
      test tag3
    </li>
  </ul>
</div>
`;
